docs/ldml/tr35-keyboards.md

```diff
@@ -308,6 +308,279 @@ For extensibility, the `<special>` element will be allowed at nearly every level
+
+**Example 3**
+
+- `e\u{0300}\m{marker1}\u{0320}a\u{0300}\m{marker2}\u{0320}` (original)
+- `e\m{marker1}\u{0320}\u{0300}a\m{marker2}\u{0320}\u{0300}` (NFD)
+
+There are two normalization-safe segments here:
+
+1. `e\u{0300}\m{marker1}\u{0320}`
+2. `a\u{0300}\m{marker2}\u{0320}`
+
+Normalization (and marker rearranging) effectively occurs within each segment.  While `\m{marker1}` is 'glued' to the `\u{0320}`, it is glued within the first segment and has no effect on the second segment.
+
+### Normalization and Character Classes
+
+If pre-composed (non-NFD) characters are used in [character classes](#regex-like-syntax), such as `[á-é]`, these may not match as keyboard authors expect, as the U+00E1 character (á) will not occur in NFD form. Thus this may be masking serious errors in the data.
+
+By default, tools that process the keyboard data should raise an error when character classes include non-NFD characters.
+
+The above should be written instead as a regex `(á|â|ã|ä|å|æ|ç|è|é)`. Alternatively, it could be written as a set variable `<set id="Example" value="á â ã ä å æ ç è é"/>` and matched as `$[Example]`.
+
+There is another case where there is no explicit mention of a non-NFD character, but the character class could include non-NFD characters, such as the range `[\u{0020}-\u{01FF}]`. For these, the tools should raise a warning by default.
+
+### Normalization and Reorders
+
+[`reorder`](#element-reorder) elements operate on NFD codepoints.
+
+### Normalization and Output
+
+On output, text will be normalized into the form requested by that implementation, or possibly specifically requested by a particular application.
+For example, many platforms may request NFC as the output format. In such a case, all text emitted via the keyboard will be transformed into NFC.
+
+Existing text in a document will only have normalization applied within a single normalization-safe segment from the caret.  Output will not contain any markers, thus any normalization is unaffected by any markers embedded within the segment.
+
+For example, the sequence `e\m{marker}\u{300}` would be output in NFC as `è`. The marker is removed and has no effect on output.
+
+### Normalization-safe Segments
+
+For purposes of the above discussion, "normalization-safe segments" are defined as a string of codepoints which are
+
+1. already in [NFD](https://www.unicode.org/reports/tr15/#Norm_Forms), and
+2. begin with a character with [Canonical Combining Class](https://www.unicode.org/reports/tr44/#Canonical_Combining_Class_Values) of `0`.
+
+See [UAX #15 Section 9.1: Stable Code Points](https://www.unicode.org/reports/tr15/#Stable_Code_Points) for related discussion.
+Text under consideration can be segmented by locating such characters.
+
+### Disabling Normalization
+
+The attribute value `normalization="disabled"` can be used to indicate that no automatic normalization is to be applied in input, matching, or output. Using this setting should be done with caution:
```

Comment:
Why are we supplying this option, when it is so dangerous

Reply:
We're supplying this option, because some text loses data during NFC.  One example given in discussion was Hebrew cantillation marks.

Reply:
From my perspective, this is an escape valve that allows a keyboard author to match behaviour in existing input methods which do not support normalization. For example, if you want to create a Vietnamese keyboard that matches other Vietnamese input methods that produce non-normalized output.

I guess this is partly recognition that we are uncertain that normalization is going to be desirable in all circumstances. It may be that it works out that no keyboards need this, in which case we deprecate in a future version. But it feels like a big call right now to say we will always normalize in all cases without any way out of it.

I'd say it's not actually dangerous -- it doesn't result in data loss -- but does potentially make a lot more work for the keyboard developer who has to consider input normalization themselves.  but no more than was previously the case with existing input methods.

We found that it was not hard to support this flag in the Keyman implementation, so cost shouldn't be a big issue.

Reply:
Semi-related: a real life example from today of keyboard authors dealing with normalization: https://github.com/keymanapp/keyboards/pull/2564

Reply:
Ok, that makes sense.